File: backend/shared/python/encryption/examples/dynamodb_integration.py
"""
DynamoDB Integration Examples for VaidyaLink Data Models
Demonstrates field-level encryption for Patient, ScanJobs, and VoiceJobs tables
"""

import os
import json
from datetime import datetime
from typing import Dict, Any, Optional, List
import boto3
from encryption import FieldEncryption


# Initialize DynamoDB client
dynamodb = boto3.resource('dynamodb', region_name=os.environ.get('AWS_REGION', 'ap-south-1'))

# Initialize encryption
field_encryption = FieldEncryption()


# Patient Table Operations

def store_patient(patient_data: Dict[str, Any]) -> Dict[str, Any]:
    """
    Store encrypted patient record in DynamoDB

    Args:
        patient_data: Patient information including sensitive PHI fields

    Returns:
        Success status and patient ID
    """
    try:
        patient_id = patient_data['patientId']

        # Encrypt sensitive fields
        encrypted_patient = field_encryption.encrypt_record(
            patient_data,
            patient_id=patient_id,
            additional_context={
                'table': 'Patients',
                'operation': 'create'
            }
        )

        # Add DynamoDB keys
        item = {
            'PK': f'PATIENT#{patient_id}',
            'SK': 'PROFILE',
            **encrypted_patient,
            'createdAt': datetime.utcnow().isoformat(),
            'updatedAt': datetime.utcnow().isoformat()
        }

        table = dynamodb.Table(os.environ.get('PATIENTS_TABLE', 'VaidyaLink-Patients'))
        table.put_item(Item=item)

        print(f'Patient {patient_id} stored with encrypted PHI fields')
        return {'success': True, 'patientId': patient_id}

    except Exception as e:
        print(f'Error storing patient: {e}')
        raise


def get_patient(patient_id: str, fields_to_decrypt: Optional[List[str]] = None) -> Optional[Dict[str, Any]]:
    """
    Retrieve and decrypt patient record from DynamoDB

    Args:
        patient_id: Patient identifier
        fields_to_decrypt: Optional list of specific fields to decrypt (decrypts all if None)

    Returns:
        Decrypted patient record or None if not found
    """
    try:
        table = dynamodb.Table(os.environ.get('PATIENTS_TABLE', 'VaidyaLink-Patients'))

        response = table.get_item(
            Key={
                'PK': f'PATIENT#{patient_id}',
                'SK': 'PROFILE'
            }
        )

        if 'Item' not in response:
            return None

        # Decrypt sensitive fields (partial or full)
        decrypted_patient = field_encryption.decrypt_record(
            response['Item'],
            patient_id=patient_id,
            additional_context={'table': 'Patients', 'operation': 'read'},
            fields_to_decrypt=fields_to_decrypt
        )

        return decrypted_patient

    except Exception as e:
        print(f'Error retrieving patient: {e}')
        raise


def update_patient(patient_id: str, updates: Dict[str, Any]) -> Dict[str, Any]:
    """
    Update patient record with encrypted fields

    Args:
        patient_id: Patient identifier
        updates: Dictionary of fields to update

    Returns:
        Success status and patient ID
    """
    try:
        # Encrypt any sensitive fields in updates
        encrypted_updates = field_encryption.encrypt_record(
            updates,
            patient_id=patient_id,
            additional_context={'table': 'Patients', 'operation': 'update'}
        )

        # Build update expression
        update_expressions = []
        expression_attribute_names = {}
        expression_attribute_values = {}

        for idx, (key, value) in enumerate(encrypted_updates.items()):
            attr_name = f'#attr{idx}'
            attr_value = f':val{idx}'
            update_expressions.append(f'{attr_name} = {attr_value}')
            expression_attribute_names[attr_name] = key
            expression_attribute_values[attr_value] = value

        # Add updatedAt
        update_expressions.append('#updatedAt = :updatedAt')
        expression_attribute_names['#updatedAt'] = 'updatedAt'
        expression_attribute_values[':updatedAt'] = datetime.utcnow().isoformat()

        table = dynamodb.Table(os.environ.get('PATIENTS_TABLE', 'VaidyaLink-Patients'))
        table.update_item(
            Key={
                'PK': f'PATIENT#{patient_id}',
                'SK': 'PROFILE'
            },
            UpdateExpression=f"SET {', '.join(update_expressions)}",
            ExpressionAttributeNames=expression_attribute_names,
            ExpressionAttributeValues=expression_attribute_values
        )

        print(f'Patient {patient_id} updated with encrypted fields')
        return {'success': True, 'patientId': patient_id}

    except Exception as e:
        print(f'Error updating patient: {e}')
        raise


# ScanJobs Table Operations

def store_scan_job(scan_job_data: Dict[str, Any]) -> Dict[str, Any]:
    """
    Store encrypted scan job with extracted data

    Args:
        scan_job_data: Scan job information including extracted medical data

    Returns:
        Success status and job ID
    """
    try:
        job_id = scan_job_data['jobId']
        patient_id = scan_job_data['patientId']

        # Encrypt sensitive extracted data
        encrypted_job = field_encryption.encrypt_record(
            scan_job_data,
            patient_id=patient_id,
            additional_context={
                'table': 'ScanJobs',
                'job_id': job_id,
                'operation': 'create'
            }
        )

        item = {
            'PK': f'JOB#{job_id}',
            'SK': 'METADATA',
            **encrypted_job,
            'createdAt': datetime.utcnow().isoformat(),
            'updatedAt': datetime.utcnow().isoformat()
        }

        table = dynamodb.Table(os.environ.get('SCANJOBS_TABLE', 'VaidyaLink-ScanJobs'))
        table.put_item(Item=item)

        print(f'Scan job {job_id} stored with encrypted extracted data')
        return {'success': True, 'jobId': job_id}

    except Exception as e:
        print(f'Error storing scan job: {e}')
        raise


def get_scan_job(job_id: str, patient_id: str) -> Optional[Dict[str, Any]]:
    """
    Retrieve and decrypt scan job

    Args:
        job_id: Scan job identifier
        patient_id: Patient identifier (required for decryption context)

    Returns:
        Decrypted scan job or None if not found
    """
    try:
        table = dynamodb.Table(os.environ.get('SCANJOBS_TABLE', 'VaidyaLink-ScanJobs'))

        response = table.get_item(
            Key={
                'PK': f'JOB#{job_id}',
                'SK': 'METADATA'
            }
        )

        if 'Item' not in response:
            return None

        # Decrypt extracted data
        decrypted_job = field_encryption.decrypt_record(
            response['Item'],
            patient_id=patient_id,
            additional_context={
                'table': 'ScanJobs',
                'job_id': job_id,
                'operation': 'read'
            }
        )

        return decrypted_job

    except Exception as e:
        print(f'Error retrieving scan job: {e}')
        raise


# VoiceJobs Table Operations

def store_voice_job(voice_job_data: Dict[str, Any]) -> Dict[str, Any]:
    """
    Store encrypted voice job with transcription

    Args:
        voice_job_data: Voice job information including transcription

    Returns:
        Success status and job ID
    """
    try:
        job_id = voice_job_data['jobId']
        patient_id = voice_job_data['patientId']

        # Encrypt transcription and extracted medical data
        encrypted_job = field_encryption.encrypt_record(
            voice_job_data,
            patient_id=patient_id,
            additional_context={
                'table': 'VoiceJobs',
                'job_id': job_id,
                'operation': 'create'
            }
        )

        item = {
            'PK': f'VOICE#{job_id}',
            'SK': 'METADATA',
            **encrypted_job,
            'createdAt': datetime.utcnow().isoformat(),
            'updatedAt': datetime.utcnow().isoformat()
        }

        table = dynamodb.Table(os.environ.get('VOICEJOBS_TABLE', 'VaidyaLink-VoiceJobs'))
        table.put_item(Item=item)

        print(f'Voice job {job_id} stored with encrypted transcription')
        return {'success': True, 'jobId': job_id}

    except Exception as e:
        print(f'Error storing voice job: {e}')
        raise


def get_voice_job(job_id: str, patient_id: str) -> Optional[Dict[str, Any]]:
    """
    Retrieve and decrypt voice job

    Args:
        job_id: Voice job identifier
        patient_id: Patient identifier (required for decryption context)

    Returns:
        Decrypted voice job or None if not found
    """
    try:
        table = dynamodb.Table(os.environ.get('VOICEJOBS_TABLE', 'VaidyaLink-VoiceJobs'))

        response = table.get_item(
            Key={
                'PK': f'VOICE#{job_id}',
                'SK': 'METADATA'
            }
        )

        if 'Item' not in response:
            return None

        # Decrypt transcription
        decrypted_job = field_encryption.decrypt_record(
            response['Item'],
            patient_id=patient_id,
            additional_context={
                'table': 'VoiceJobs',
                'job_id': job_id,
                'operation': 'read'
            }
        )

        return decrypted_job

    except Exception as e:
        print(f'Error retrieving voice job: {e}')
        raise


# Example: Complete workflow for document processing

def complete_document_processing_workflow():
    """
    Example workflow demonstrating complete document processing with encryption
    """
    try:
        # 1. Create patient record
        patient_data = {
            'patientId': 'patient-123',
            'name': 'Rajesh Kumar',
            'dateOfBirth': '1985-06-15',
            'phone': '+91-9876543210',
            'email': 'rajesh.kumar@example.com',
            'abhaId': '12-3456-7890-1234',
            'address': '123 MG Road, Mumbai, Maharashtra',
            'preferredLanguage': 'hi'
        }

        store_patient(patient_data)

        # 2. Process document scan
        scan_job_data = {
            'jobId': 'scan-456',
            'patientId': 'patient-123',
            'status': 'completed',
            'imageS3Key': 'raw/patient-123/scan-456/original.jpg',
            'extractedData': json.dumps({
                'diagnosis': 'Type 2 Diabetes Mellitus',
                'prescriptionDetails': 'Metformin 500mg twice daily',
                'doctorNotes': 'Patient shows good compliance with medication'
            }),
            'confidenceScores': {
                'diagnosis': 0.95,
                'prescriptionDetails': 0.92
            }
        }

        store_scan_job(scan_job_data)

        # 3. Process voice recording
        voice_job_data = {
            'jobId': 'voice-789',
            'patientId': 'patient-123',
            'status': 'completed',
            'audioS3Key': 'audio/patient-123/voice-789/recording.wav',
            'language': 'hi',
            'transcription': 'मुझे पिछले तीन महीनों से सिरदर्द हो रहा है',
            'transcribedText': 'I have been having headaches for the past three months',
            'transcriptionConfidence': 0.89
        }

        store_voice_job(voice_job_data)

        # 4. Retrieve patient with only name decrypted (for display)
        patient_summary = get_patient('patient-123', fields_to_decrypt=['name'])
        print(f'Patient summary: {patient_summary}')

        # 5. Retrieve full scan job data
        scan_job = get_scan_job('scan-456', 'patient-123')
        print(f'Scan job with decrypted data: {scan_job}')

        print('Complete workflow executed successfully')

    except Exception as e:
        print(f'Workflow error: {e}')
        raise


if __name__ == '__main__':
    # Run example workflow
    complete_document_processing_workflow()

File: backend/shared/nodejs/encryption/examples/dynamodb-integration.js
/**
 * DynamoDB Integration Examples for VaidyaLink Data Models
 * Demonstrates field-level encryption for Patient, ScanJobs, and VoiceJobs tables
 */

const { DynamoDBClient } = require('@aws-sdk/client-dynamodb');
const {
  DynamoDBDocumentClient,
  PutCommand,
  GetCommand,
  UpdateCommand,
} = require('@aws-sdk/lib-dynamodb');
const { FieldEncryption } = require('../field-encryption');

// Initialize DynamoDB client
const dynamoClient = new DynamoDBClient({ region: process.env.AWS_REGION || 'ap-south-1' });
const docClient = DynamoDBDocumentClient.from(dynamoClient);

// Initialize encryption
const fieldEncryption = new FieldEncryption();

/**
 * Patient Table Operations
 */

/**
 * Store encrypted patient record in DynamoDB
 */
async function storePatient(patientData) {
  try {
    const patientId = patientData.patientId;

    // Encrypt sensitive fields
    const encryptedPatient = await fieldEncryption.encryptRecord(patientData, patientId, {
      table: 'Patients',
      operation: 'create',
    });

    // Add DynamoDB keys
    const item = {
      PK: `PATIENT#${patientId}`,
      SK: 'PROFILE',
      ...encryptedPatient,
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    };

    await docClient.send(
      new PutCommand({
        TableName: process.env.PATIENTS_TABLE || 'VaidyaLink-Patients',
        Item: item,
      })
    );

    console.log(`Patient ${patientId} stored with encrypted PHI fields`);
    return { success: true, patientId };
  } catch (error) {
    console.error('Error storing patient:', error);
    throw error;
  }
}

/**
 * Retrieve and decrypt patient record from DynamoDB
 */
async function getPatient(patientId, fieldsToDecrypt = null) {
  try {
    const result = await docClient.send(
      new GetCommand({
        TableName: process.env.PATIENTS_TABLE || 'VaidyaLink-Patients',
        Key: {
          PK: `PATIENT#${patientId}`,
          SK: 'PROFILE',
        },
      })
    );

    if (!result.Item) {
      return null;
    }

    // Decrypt sensitive fields (partial or full)
    const decryptedPatient = await fieldEncryption.decryptRecord(
      result.Item,
      patientId,
      { table: 'Patients', operation: 'read' },
      fieldsToDecrypt
    );

    return decryptedPatient;
  } catch (error) {
    console.error('Error retrieving patient:', error);
    throw error;
  }
}

/**
 * Update patient record with encrypted fields
 */
async function updatePatient(patientId, updates) {
  try {
    // Encrypt any sensitive fields in updates
    const encryptedUpdates = await fieldEncryption.encryptRecord(updates, patientId, {
      table: 'Patients',
      operation: 'update',
    });

    // Build update expression
    const updateExpressions = [];
    const expressionAttributeNames = {};
    const expressionAttributeValues = {};

    Object.entries(encryptedUpdates).forEach(([key, value], index) => {
      const attrName = `#attr${index}`;
      const attrValue = `:val${index}`;
      updateExpressions.push(`${attrName} = ${attrValue}`);
      expressionAttributeNames[attrName] = key;
      expressionAttributeValues[attrValue] = value;
    });

    // Add updatedAt
    updateExpressions.push('#updatedAt = :updatedAt');
    expressionAttributeNames['#updatedAt'] = 'updatedAt';
    expressionAttributeValues[':updatedAt'] = new Date().toISOString();

    await docClient.send(
      new UpdateCommand({
        TableName: process.env.PATIENTS_TABLE || 'VaidyaLink-Patients',
        Key: {
          PK: `PATIENT#${patientId}`,
          SK: 'PROFILE',
        },
        UpdateExpression: `SET ${updateExpressions.join(', ')}`,
        ExpressionAttributeNames: expressionAttributeNames,
        ExpressionAttributeValues: expressionAttributeValues,
      })
    );

    console.log(`Patient ${patientId} updated with encrypted fields`);
    return { success: true, patientId };
  } catch (error) {
    console.error('Error updating patient:', error);
    throw error;
  }
}

/**
 * ScanJobs Table Operations
 */

/**
 * Store encrypted scan job with extracted data
 */
async function storeScanJob(scanJobData) {
  try {
    const jobId = scanJobData.jobId;
    const patientId = scanJobData.patientId;

    // Encrypt sensitive extracted data
    const encryptedJob = await fieldEncryption.encryptRecord(scanJobData, patientId, {
      table: 'ScanJobs',
      job_id: jobId,
      operation: 'create',
    });

    const item = {
      PK: `JOB#${jobId}`,
      SK: 'METADATA',
      ...encryptedJob,
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    };

    await docClient.send(
      new PutCommand({
        TableName: process.env.SCANJOBS_TABLE || 'VaidyaLink-ScanJobs',
        Item: item,
      })
    );

    console.log(`Scan job ${jobId} stored with encrypted extracted data`);
    return { success: true, jobId };
  } catch (error) {
    console.error('Error storing scan job:', error);
    throw error;
  }
}

/**
 * Retrieve and decrypt scan job
 */
async function getScanJob(jobId, patientId) {
  try {
    const result = await docClient.send(
      new GetCommand({
        TableName: process.env.SCANJOBS_TABLE || 'VaidyaLink-ScanJobs',
        Key: {
          PK: `JOB#${jobId}`,
          SK: 'METADATA',
        },
      })
    );

    if (!result.Item) {
      return null;
    }

    // Decrypt extracted data
    const decryptedJob = await fieldEncryption.decryptRecord(result.Item, patientId, {
      table: 'ScanJobs',
      job_id: jobId,
      operation: 'read',
    });

    return decryptedJob;
  } catch (error) {
    console.error('Error retrieving scan job:', error);
    throw error;
  }
}

/**
 * VoiceJobs Table Operations
 */

/**
 * Store encrypted voice job with transcription
 */
async function storeVoiceJob(voiceJobData) {
  try {
    const jobId = voiceJobData.jobId;
    const patientId = voiceJobData.patientId;

    // Encrypt transcription and extracted medical data
    const encryptedJob = await fieldEncryption.encryptRecord(voiceJobData, patientId, {
      table: 'VoiceJobs',
      job_id: jobId,
      operation: 'create',
    });

    const item = {
      PK: `VOICE#${jobId}`,
      SK: 'METADATA',
      ...encryptedJob,
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    };

    await docClient.send(
      new PutCommand({
        TableName: process.env.VOICEJOBS_TABLE || 'VaidyaLink-VoiceJobs',
        Item: item,
      })
    );

    console.log(`Voice job ${jobId} stored with encrypted transcription`);
    return { success: true, jobId };
  } catch (error) {
    console.error('Error storing voice job:', error);
    throw error;
  }
}

/**
 * Retrieve and decrypt voice job
 */
async function getVoiceJob(jobId, patientId) {
  try {
    const result = await docClient.send(
      new GetCommand({
        TableName: process.env.VOICEJOBS_TABLE || 'VaidyaLink-VoiceJobs',
        Key: {
          PK: `VOICE#${jobId}`,
          SK: 'METADATA',
        },
      })
    );

    if (!result.Item) {
      return null;
    }

    // Decrypt transcription
    const decryptedJob = await fieldEncryption.decryptRecord(result.Item, patientId, {
      table: 'VoiceJobs',
      job_id: jobId,
      operation: 'read',
    });

    return decryptedJob;
  } catch (error) {
    console.error('Error retrieving voice job:', error);
    throw error;
  }
}

/**
 * Example: Complete workflow for document processing
 */
async function completeDocumentProcessingWorkflow() {
  try {
    // 1. Create patient record
    const patientData = {
      patientId: 'patient-123',
      name: 'Rajesh Kumar',
      dateOfBirth: '1985-06-15',
      phone: '+91-9876543210',
      email: 'rajesh.kumar@example.com',
      abhaId: '12-3456-7890-1234',
      address: '123 MG Road, Mumbai, Maharashtra',
      preferredLanguage: 'hi',
    };

    await storePatient(patientData);

    // 2. Process document scan
    const scanJobData = {
      jobId: 'scan-456',
      patientId: 'patient-123',
      status: 'completed',
      imageS3Key: 'raw/patient-123/scan-456/original.jpg',
      extractedData: JSON.stringify({
        diagnosis: 'Type 2 Diabetes Mellitus',
        prescriptionDetails: 'Metformin 500mg twice daily',
        doctorNotes: 'Patient shows good compliance with medication',
      }),
      confidenceScores: {
        diagnosis: 0.95,
        prescriptionDetails: 0.92,
      },
    };

    await storeScanJob(scanJobData);

    // 3. Process voice recording
    const voiceJobData = {
      jobId: 'voice-789',
      patientId: 'patient-123',
      status: 'completed',
      audioS3Key: 'audio/patient-123/voice-789/recording.wav',
      language: 'hi',
      transcription: 'मुझे पिछले तीन महीनों से सिरदर्द हो रहा है',
      transcribedText: 'I have been having headaches for the past three months',
      transcriptionConfidence: 0.89,
    };

    await storeVoiceJob(voiceJobData);

    // 4. Retrieve patient with only name decrypted (for display)
    const patientSummary = await getPatient('patient-123', ['name']);
    console.log('Patient summary:', patientSummary);

    // 5. Retrieve full scan job data
    const scanJob = await getScanJob('scan-456', 'patient-123');
    console.log('Scan job with decrypted data:', scanJob);

    console.log('Complete workflow executed successfully');
  } catch (error) {
    console.error('Workflow error:', error);
    throw error;
  }
}

module.exports = {
  storePatient,
  getPatient,
  updatePatient,
  storeScanJob,
  getScanJob,
  storeVoiceJob,
  getVoiceJob,
  completeDocumentProcessingWorkflow,
};
